services/galley/src/Galley/API/MLS/Message.hs

```diff
@@ -1123,53 +1124,77 @@ executeProposalAction qusr con lconv mlsMeta cm action = do
-      -- get list of mls clients from brig
-      clientInfo <- getClientInfo lconv qtarget ss
-      let allClients = Set.map ciId clientInfo
-      let allMLSClients = Set.map ciId (Set.filter ciMLS clientInfo)
-      -- We check the following condition:
-      --   allMLSClients ⊆ clients ⊆ allClients
-      -- i.e.
-      -- - if a client has at least 1 key package, it has to be added
-      -- - if a client is being added, it has to still exist
-      --
-      -- The reason why we can't simply check that clients == allMLSClients is
-      -- that a client with no remaining key packages might be added by a user
-      -- who just fetched its last key package.
-      unless
-        ( Set.isSubsetOf allMLSClients clients
-            && Set.isSubsetOf clients allClients
-        )
-        $ do
-          -- unless (Set.isSubsetOf allClients clients) $ do
-          -- FUTUREWORK: turn this error into a proper response
-          throwS @'MLSClientMismatch
+  failedAddFetching <- fmap catMaybes $
+    forM newUserClients $
+      \(qtarget, newclients) -> case Map.lookup qtarget cm of
+        -- user is already present, skip check in this case
+        Just _ -> do
+          -- new user
+          pure Nothing
+        Nothing -> do
+          -- final set of clients in the conversation
+          let clients = Set.map fst (newclients <> Map.findWithDefault mempty qtarget cm)
+          -- get list of mls clients from Brig (local or remote)
+          getClientInfo lconv qtarget ss >>= \case
+            Left _e -> pure (Just qtarget)
+            Right clientInfo -> do
+              let allClients = Set.map ciId clientInfo
+              let allMLSClients = Set.map ciId (Set.filter ciMLS clientInfo)
+              -- We check the following condition:
+              --   allMLSClients ⊆ clients ⊆ allClients
+              -- i.e.
+              -- - if a client has at least 1 key package, it has to be added
+              -- - if a client is being added, it has to still exist
+              --
+              -- The reason why we can't simply check that clients == allMLSClients is
+              -- that a client with no remaining key packages might be added by a user
+              -- who just fetched its last key package.
+              unless
+                ( Set.isSubsetOf allMLSClients clients
+                    && Set.isSubsetOf clients allClients
+                )
+                $ do
+                  -- FUTUREWORK: turn this error into a proper response
+                  throwS @'MLSClientMismatch
+              pure Nothing
 
   membersToRemove <- catMaybes <$> for removedUsers (uncurry checkRemoval)
 
   -- add users to the conversation and send events
-  addEvents <- foldMap addMembers . nonEmpty . map fst $ newUserClients
+  addEvents <-
+    foldMap addMembers
+      . nonEmpty
+      . filter (\u -> u `notElem` failedAddFetching)
+      . fmap fst
+      $ newUserClients
+  let failedAdding =
+        Set.toList $
```

Comment:
I thought it would be much simpler yesterday when we looked at this, but I couldn't make it so. It is a bit simpler now, though. See commit fc010ea3e.